### ISSUE TITLE
consider supporting alternate options for subnet discovery

### ISSUE THREAD

**Issue by trondhindenes** [awslabs/karpenter#632]
**Tell us about your request**
I noticed that Karpenter relies on vpc subnet tags in order to discover subnets. This may be problematic in situations where a single vpc hosts a number of kubernetes clusters, and the vpc is provisioned by a different process than each cluster. Having to tag subnets with each cluster's name breaks the notion of clean separation between networking and compute that many orgs are trying to achieve when designing their IAC setups.

Therefore, I would vote for allowing karpenter to discover subnets using an alternative method such as a config map, a custom crd or similar. In cases where EKS EniConfig objects exist (https://docs.aws.amazon.com/eks/latest/userguide/cni-custom-network.html) it might make sense to piggyback off of those.

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
Tried to explain it above

**Are you currently working around this issue?**
Not currently using Karpenter, looking to start testing it

**Additional context**

**Attachments**

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by ellistarn** [comment]
Hey @trondhindenes. We currently support `node.k8s.aws/subnet-tag-key` (currently docs are behind the implementation) which will select any subnet with the specified key. Does this meet your usecase, or are you looking to specify multiple subnet names?

**Comment by trondhindenes** [comment]
hi @ellistarn Could you explain a bit more? Where is `node.k8s.aws/subnet-tag-key` used?

**Comment by trondhindenes** [comment]
Digging a bit more now - looks like the label you mention is part of the `constraints` object of the `Provisioner`, but can be optionally overridden by pod-level labelling.

If that's correct then I think this will work perfectly and is a good alternative to subnet tagging. I'll try and test this out today!

**Comment by trondhindenes** [comment]
I got stuck pretty quickly because of https://github.com/awslabs/karpenter/issues/646 - wasn't able to add the `node.k8s.aws/subnet-tag-key` to the provisioner object :-(

**Comment by ellistarn** [comment]
> Digging a bit more now - looks like the label you mention is part of the `constraints` object of the `Provisioner`, but can be optionally overridden by pod-level labelling.
> 
> If that's correct then I think this will work perfectly and is a good alternative to subnet tagging. I'll try and test this out today!

We desperately need to document this.

**Comment by trondhindenes** [comment]
While working to troubleshoot #646 to answer your question: We definetely need to be able to specify multiple subnets. My VPC is provisioned using CDK, and by default there's no way to reliably indicate a subnet group using the above-mentioned subnets.

**Comment by ellistarn** [comment]
Hey @trondhindenes ,

I'm exploring this right now:
```

// Extensions are parameters specific to this cloud provider
// +kubebuilder:object:root=true
type AWS struct {
    ...
	Subnets []string `json:"subnets,omitempty"`
	// SubnetSelector discovers subnets by tags. A value of "" is a wildcard.
	// +optional
	SubnetSelector map[string]string `json:"subnetSelector,omitempty"`
	// SecurityGroups specify the names of the security groups.
	// +optional
	SecurityGroups []string `json:"securityGroups,omitempty"`
	// SecurityGroupSelector discovers security groups by tags. A value of "" is a wildcard.
	// +optional
	SecurityGroupsSelector map[string]string `json:"securityGroupSelector,omitempty"`
}
```
Do you have any preference for which mechanism you'd use? We can include both, but I was considering starting with just the selectors.

**Comment by trondhindenes** [comment]
thanks @ellistarn ,
I'm afraid I don't quite understand the question - are you asking if we prefer to to include subnets based on a key or if we prefer key+value? In order for us to be able to use subnets provisioned by CDK, I think we need the latter (and the tag value should support the `/` character and upper/lower-casing etc.)

**Comment by bwagner5** [comment]
Implemented in v1alpha4 `v0.3.4`
